File: app/src/main/java/com/waveneuro/domain/usecase/login/LoginUseCase.java
package com.waveneuro.domain.usecase.login;

import com.asif.abase.domain.base.ObservableUseCase;
import com.asif.abase.domain.base.UseCaseCallback;
import com.waveneuro.data.DataManager;
import com.waveneuro.data.model.request.login.LoginRequest;
import com.waveneuro.data.model.response.login.LoginResponse;
import com.waveneuro.data.model.response.login.LoginResponseMfa;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Observable;

public class LoginUseCase extends ObservableUseCase<LoginResponseMfa> {

    private final DataManager dataManager;

    private LoginRequest request;

    @Inject
    public LoginUseCase(DataManager dataManager) {
        this.dataManager = dataManager;
    }

    @Override
    public Observable<LoginResponseMfa> buildUseCaseSingle() {
        return dataManager.login(request);
    }

    public void execute(LoginRequest request, UseCaseCallback<LoginResponseMfa> useCaseCallback) {
        this.request=request;
        super.execute(useCaseCallback);
    }

}

File: app/src/main/java/com/waveneuro/domain/usecase/login/LoginUseCase.kt
package com.waveneuro.domain.usecase.login

import com.waveneuro.data.api.user.UserApi
import com.waveneuro.data.api.user.model.login.ApiLoginRq
import com.waveneuro.domain.model.login.LoginMapperImpl
import com.waveneuro.domain.model.login.LoginMfaRs
import javax.inject.Inject

class LoginUseCase @Inject constructor(
    private val userApi: UserApi,
    private val mapper: LoginMapperImpl
) {

    suspend fun login(username: String, password: String): LoginMfaRs =
        mapper.fromApiToDomain(userApi.login(ApiLoginRq(username, password)))

}